src/backend/distributed/transaction/remote_transaction.c

```diff
@@ -115,25 +115,41 @@ void
 FinishRemoteTransactionBegin(struct MultiConnection *connection)
 {
 	RemoteTransaction *transaction = &connection->remoteTransaction;
-	PGresult *result = NULL;
 	const bool raiseErrors = true;
+	bool failed = false;
 
 	Assert(transaction->transactionState == REMOTE_TRANS_STARTING);
 
-	result = GetRemoteCommandResult(connection, raiseErrors);
-	if (!IsResponseOK(result))
-	{
-		ReportResultError(connection, result, WARNING);
-		MarkRemoteTransactionFailed(connection, raiseErrors);
-	}
-	else
+	/*
+	 * We consume all command results and report the first
+	 * failure if there is any. There is no need to call
+	 * ForgetResults() after this.
+	 */
+	while (true)
 	{
-		transaction->transactionState = REMOTE_TRANS_STARTED;
-		transaction->lastSuccessfulSubXact = transaction->lastQueuedSubXact;
-	}
+		PGresult *result = GetRemoteCommandResult(connection, raiseErrors);
 
-	PQclear(result);
-	ForgetResults(connection);
+		/* all results consumed, no failure encountered */
+		if (result == NULL && !failed)
+		{
+			transaction->transactionState = REMOTE_TRANS_STARTED;
+			transaction->lastSuccessfulSubXact = transaction->lastQueuedSubXact;
+			break;
+		}
+		else if (result == NULL)
+		{
+			break;
+		}
+
+		if (!IsResponseOK(result) && !failed)
```

Comment:
I think it would be better to move the error checks in `ForgetResults`, since there doesn't seem to be a case in which we wouldn't want to check for failure there. We may have to add a `raiseErrors` flag.

Reply:
per our discussion, will make `ForgetResults()` to return boolean to indicate if anything went wrong when forgetting.